### PR TITLE
[resolves #2595][ENTESB-7959] Support securing Fuse CXF endpoints with Elytron

### LINKED ISSUE #2595: Support securing Fuse CXF endpoints with Elytron
Elytron is a new Security provider on WildFly. We want to reach the state when it is possible to secure CXF endpoints through Elytron security domains. See also https://issues.jboss.org/browse/ENTESB-7959

### PR DESCRIPTION
**Pull request by ppalaga**
I am sending this against `camel-2.22.x` branch because this PR depends on the `CAMEL-12632: Pass CXF service class to EndpointInfo` commit available there. I could successfully build with Camel revision https://github.com/apache/camel/commit/7f7896b673db1f471528350bccab0b808eb6c6f3

About this PR:

* I was able to solve the "CDI in WAR in EAR" case (and there is a test for it).
* This PR introduces the following limitation: Service classes of CXF endpoints need to be defined inside a WAR. In the past, it was possible to deploy a such an enpoint inside a jar. That does not work anymore. I hope that's acceptable. I had to change existing CXF itests to use .war deployments instead of .jars. I'd squash the separate commit once we agree this is OK. There are two technical reasons behind that: (1) my solution for the "CDI in WAR in EAR" case uses the service class to find the WAR class loader. (2) We clone and adapt the deploymentInfo of the main deployment to get the security filters set up properly and the main deploymentInfo is there only for WARs.
* With current WildFly, I have not found a way to tell for a given security domain name whether it is legacy or Elytron inside DUP.deploy(). As a consequence of that, I could not make the Elytron secured endpoints fail when their path is outside the main deployment's context path. I could only make all out of context endpoints fail or none. Making fail all would be backwards incompatibe. So I decided to keep supporting Elytron also for out of context endpoints via the double slash web.xml convention.

### CODE REVIEW

File: cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java [hunk 1]
```diff
@@ -138,7 +129,14 @@ public String getAddress(EndpointInfo endpointInfo) {
      */
     public void activate() {
         super.activate();
-        LOG.log(Level.FINE, "Activating receipt of incoming messages");
+        LOG.debug("Activating receipt of incoming messages");
+
+        Class<?> serviceClass = endpointInfo.getProperty("serviceClass", Class.class);
+        if (serviceClass != null && this.loader == null) {
+            final ClassLoader cl = serviceClass.getClassLoader();
```
[the remark this answers is not in the document]
Comment (jamesnetherton) [reply cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java]: Also, maybe a bit hacky, but could we perhaps get the `ModuleIdentifier` to determine whether we're dealing with a JAR or WAR deployment? I didn't seen any handling for that scenario elsewhere. E.g something like:

```
ModuleIdentifier moduleId = cl.getModule().getIdentifier();
if (moduleId.getName().endsWith(".jar") {
  throw new RuntimeException("CXF endpoints can only be bootstrapped via WAR deployments");
}
```

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java [hunk 2]
```diff
@@ -57,12 +59,15 @@
...
     public static final int DEPENDENCIES_CAMEL_INTEGRATION = Phase.DEPENDENCIES_LOGGING + 0x01;
     public static final int DEPENDENCIES_CAMEL_WIRINGS = DEPENDENCIES_CAMEL_INTEGRATION + 0x01;
+    public static final int DEPENDENCIES_CAMEL_ENDPOINT_DEPLOYMENT_SCHEDULER = DEPENDENCIES_CAMEL_WIRINGS + 0x01;
 
     public static final int INSTALL_PACKAGE_SCAN_RESOLVER = Phase.INSTALL_WS_DEPLOYMENT_ASPECTS + 0x01;
     public static final int INSTALL_CDI_BEAN_ARCHIVE_PROCESSOR = INSTALL_PACKAGE_SCAN_RESOLVER + 0x01;
     public static final int INSTALL_CAMEL_CONTEXT_CREATE = INSTALL_CDI_BEAN_ARCHIVE_PROCESSOR + 0x01;
     public static final int INSTALL_CONTEXT_ACTIVATION = INSTALL_CAMEL_CONTEXT_CREATE + 0x01;
 
+    public static final int INSTALL_CAMEL_ENDPOINT_DEPLOYER = Phase.INSTALL_WAR_DEPLOYMENT + 0x01;
```
Comment (jamesnetherton) [inline subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: What's not clear to me yet, is the consequence of having the endpoint deployer run after the Camel context has been started (I assume that's what happens?).

E.g what happens if I have a route like:

```
from("timer:foo?period=1s&delay=0")
.to("cxfrs://http://foo/bar?resourceClasses=org.foo.bar")
```

Does the Camel context get fully started before the CXF endpoint is deployed. I.e is there some period where the above code would be polling an endpoint that is not yet available?

Maybe it's not an issue (not actually sure whether the current solution deals with this), but it'd be good to check.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+            } else if (result.charAt(0) != '/') {
+                return "/" + result.toString();
+            } else {
+                return result.toString();
+            }
+        }
+
+    }
+
+    private static final String CATCH_ALL_ENDPOINT_URI_PREFIX = "///*";
```
Comment (jamesnetherton) [inline subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: Can we keep constants at the top of the class please.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+    private static final String CATCH_ALL_PREFIX = "/*";
```
Comment (jamesnetherton) [inline subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: Not keen on the `CATCH_ALL` prefix but can't think of an decent alternative at the moment.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+    private static final String SERVICE_NAME = "EndpointDeployer";
+
+    /**
+     * This method can simplified substantially, once https://github.com/undertow-io/undertow/pull/642 reaches us.
+     * Currently, the method is just an adjusted copy of {@link DeploymentInfo#clone()}.
+     *
+     * @param securityConstraints
+     * @param servletInfo
+     * @param deploymentName
+     * @param contextPath
```
Comment (jamesnetherton) [inline subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: The `@param` documentation does not seem to be correct for the method signature.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                .install();
+    }
+
+    public static ServiceName deploymentSchedulerServiceName(ClassLoader deploymentClassLoader) {
+        if (deploymentClassLoader == null) {
+            deploymentClassLoader = SecurityActions.getContextClassLoader();
+        }
+        if (deploymentClassLoader instanceof ModuleClassLoader) {
+            ModuleClassLoader moduleClassLoader = (ModuleClassLoader) deploymentClassLoader;
+            final String DEPLOYMEND_CL_NAME_PREFIX = "deployment.";
```
Comment (jamesnetherton) [inline subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: Typo `DEPLOYMEND`.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    public CamelEndpointDeploymentSchedulerService getValue() throws IllegalStateException {
+        return this;
+    }
+
+    /**
+     * Either schedules the given HTTP endpoint for deployment once {@link #deployerService} becomes available or
+     * deploys it instantly if the {@link #deployerService} is available already.
+     *
+     * @param uri
+     *                                determines the path and protocol under which the HTTP endpoint should be exposed
```
Comment (jamesnetherton) [inline subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: Noticed in a few places, the Javadoc indentation / line continuation seems a bit weird.

**Comment by jamesnetherton** [comment]
We should also check that the wildfly-camel-examples test suite passes ok with these changes.

File: cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppalaga) [reply cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java]: > Shall we check that the ClassLoader is an instance of ModuleClassLoader and throw an exception if is not? It's sort of an illegal state if we don't have the expected ClassLoader.

Done when the class loader name is transformed into a DeploymentSchedulerService name https://github.com/ppalaga/wildfly-camel/blob/8bed967d29537e095b7b699ca7e531121ae8d24b/subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java#L75

> Also, maybe a bit hacky, but could we perhaps get the ModuleIdentifier to determine whether we're dealing with a JAR or WAR deployment? I didn't seen any handling for that scenario elsewhere. E.g something like:
>
> ModuleIdentifier moduleId = cl.getModule().getIdentifier();
> if (moduleId.getName().endsWith(".jar") {
>   throw new RuntimeException("CXF endpoints can only be bootstrapped via WAR deployments");
> }

ModuleIdentifier is deprecated, but yes, such a check can be done on `ModuleClassLoader.getName()` and such a check is not there yet and thanks, that's a good idea :)

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        addHttpAuthentication.get("security-domain").set(SECURITY_DOMAIN);
+        addHttpAuthentication.get("http-server-mechanism-factory").set("global");
+        addHttpAuthentication.get("mechanism-configurations").get(0).get("mechanism-name").set("BASIC");
+        addHttpAuthentication.get("mechanism-configurations").get(0).get("mechanism-realm-configurations").get(0)
+                .get("realm-name").set(SECURITY_REALM);
+
+        final ModelNode addUndertowDomain = Util.createAddOperation(UNDERTOW_DOMAIN_ADDRESS);
+        addUndertowDomain.get("http-authentication-factory").set(HTTP_AUTH_FACTORY);
+
+        DMRUtils.batchNode() //
```
[the remark this answers is not in the document]
Comment (ppalaga) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: That's a simple and IDE independent way to force IDE formatters to put a single builder method per line. Isn't it elegant? I am proud of the invention ;)

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/cxf/ws/secure/CXFWSSecureUtils.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            + "</greet>" //
+            + "</Body>" //
+            + "</Envelope>";
+
+    static void assertGreet(Path wildFlyHome, String uri, String user, String password, int responseCode,
+            String responseBody) throws KeyManagementException, UnrecoverableKeyException, NoSuchAlgorithmException,
+            KeyStoreException, CertificateException, IOException {
+        try (CloseableHttpClient httpclient = HttpClients.custom()
+                .setSSLSocketFactory(SecurityUtils.createBasicSocketFactory(wildFlyHome)).build()) {
+            HttpPost request = new HttpPost(uri);
```
[the remark this answers is not in the document]
Comment (ppalaga) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/cxf/ws/secure/CXFWSSecureUtils.java]: I tried, but it sucked for some reason. The HttpClient dep was there anyway already.

File: subsystem/core/pom.xml [hunk 10]
```diff
@@ -74,6 +74,18 @@
             <artifactId>gravia-container-wildfly-extension</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (ppalaga) [reply subsystem/core/pom.xml]: +1

[the remark this answers is not in the document]
Comment (ppalaga) [reply subsystem/core/pom.xml]: +1

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java (hunk 2 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: > Does the Camel context get fully started before the CXF endpoint is deployed?

Yes, I think so, at least for the CDI case. Although I am not sure when exactly the timer:foo?period=1s&delay=0 would start. 

> it'd be good to check.

You mean just investigate what happens when? Or write some sort of a test?

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 3 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: Eclipse puts inner classes before static fields by default. I am ready to adhere to any other ordering if you please tell me the complete definition?

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 4 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: `MATCH_ALL...` perhaps better?

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 5 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: @jamesnetherton has eagle's eyes!

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java (hunk 6 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: +1

Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: Also made a proper constant out of it.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java (hunk 7 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: Let me check if Eclipse formatter can be persuaded not to do that.

Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: Yeah, found some settings. Is this OK?

```
     * @param uri determines the path and protocol under which the HTTP endpoint should be exposed
     * @param endpointHttpHandler an {@link EndpointHttpHandler} to use for handling HTTP requests sent to the given
     *        {@link URI}'s path
```

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java (hunk 2 above)
Comment (jamesnetherton) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: It's kind of the same thing. You can investigate by writing a test. We don't necessarily have to commit the test, but I'm interested in what the behaviour is.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 4 above)
Comment (jamesnetherton) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: `MATCH_ALL` sounds better :+1:

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java (hunk 7 above)
Comment (jamesnetherton) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: Still find it weird that it leaves random whitespace on the new line. But hey, it's an improvement..

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 3 above)
Comment (jamesnetherton) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: Ok, that's a weird convention. I hate seeing fields and static fields spread at random throughout the class. They're better placed all together at the top IMO.

Maybe we should consider introducing checkstyle or add Sonar integration to PRs.

**Comment by jamesnetherton** [comment]
The other thing we need is some documentation explaining how to set up security, what the conventions are, any limitations etc.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 4 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: OK, let me rename it

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java (hunk 3 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeployerService.java]: Strange enough, that neither checkstyle's DeclarationOrderCheck [1] nor Sun's Code Conventions for the Java Programming Language [2] say where the inner classes should be placed.

[1] http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html 

[2] http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-141855.html#1852

**Comment by ppalaga** [comment]
> The other thing we need is some documentation explaining how to set up security, what the conventions are, any limitations etc.

+1 where is the proper location for that kind of docs? (Could you perhaps reply via new issue assigned to me?)

File: subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java (hunk 7 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/service/CamelEndpointDeploymentSchedulerService.java]: > it leaves random whitespace on the new line

It is aligned with the first word after `@param`. Apparently this is the first multiline param comment in the whole WFC codebase so I dare to declare this as the new official convention :)

**Comment by ppalaga** [comment]
e262a0a + 1d2829d addresses all requests of @jamesnetherton , except for trailing `//`, "our own HTTP client", check the context vs CXF endpoint init timing, where in the outer class should inner classes be located and wildfly-camel-examples

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesnetherton) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: IntelliJ can handle the formatting of chained method calls. From what I've read Eclipse can too.

**Comment by ppalaga** [comment]
wildfly-camel-examples are passing with e262a0a + 1d2829d

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java (hunk 2 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: Here is an excerpt from a log:

```
2018-07-12 09:05:16,462 INFO  [org.apache.camel.cdi.CdiCamelExtension] (MSC service thread 1-3) Camel CDI is starting Camel context [cxfws-secure-cdi-camel-context]
2018-07-12 09:05:16,466 INFO  [org.apache.camel.impl.DefaultCamelContext] (MSC service thread 1-3) Apache Camel 2.22.1-SNAPSHOT (CamelContext: cxfws-secure-cdi-camel-context) is starting
2018-07-12 09:05:16,467 INFO  [org.apache.camel.management.ManagedManagementStrategy] (MSC service thread 1-3) JMX is enabled
2018-07-12 09:05:16,552 INFO  [org.wildfly.extension.camel] (MSC service thread 1-3) Camel context starting: cxfws-secure-cdi-camel-context
2018-07-12 09:05:16,552 INFO  [org.wildfly.extension.camel] (MSC service thread 1-4) Bound camel naming object: java:jboss/camel/context/cxfws-secure-cdi-camel-context
2018-07-12 09:05:17,547 INFO  [org.apache.cxf.wsdl.service.factory.ReflectionServiceFactoryBean] (MSC service thread 1-3) Creating Service {http://subA.secure.ws.cxf.test.camel.wildfly.org/}GreetingServiceService from class org.wildfly.camel.test.cxf.ws.secure.subA.GreetingService
2018-07-12 09:05:18,101 INFO  [org.apache.cxf.endpoint.ServerImpl] (MSC service thread 1-3) Setting the server's publish address to be https://localhost:8443/webservices/greeting-secure-cdi
2018-07-12 09:05:18,335 INFO  [org.apache.camel.impl.DefaultCamelContext] (MSC service thread 1-3) Route: route1 started and consuming from: direct://start
2018-07-12 09:05:18,339 ERROR [org.wildfly.extension.camel] (MSC service thread 1-3) --- scheduling https://localhost:8443/webservices/greeting-secure-cdi: java.lang.RuntimeException
2018-07-12 09:05:18,370 INFO  [org.apache.camel.impl.DefaultCamelContext] (MSC service thread 1-3) Total 6 routes, of which 6 are started
2018-07-12 09:05:18,371 INFO  [org.apache.camel.impl.DefaultCamelContext] (MSC service thread 1-3) Apache Camel 2.22.1-SNAPSHOT (CamelContext: cxfws-secure-cdi-camel-context) started in 1.906 seconds
2018-07-12 09:05:18,415 ERROR [org.wildfly.extension.camel] (MSC service thread 1-6) --- deploying https://localhost:8443/webservices/greeting-secure-cdi: java.lang.RuntimeException
2018-07-12 09:05:18,467 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-6) WFLYUT0021: Registered web context: '/webservices/greeting-secure-cdi' for server 'default-server'
2018-07-12 09:05:18,468 INFO  [org.wildfly.extension.camel] (MSC service thread 1-6) Add Camel endpoint: http://127.0.0.1:8080/webservices/greeting-secure-cdi
```

Note that the endpoint is exposed at `09:05:18,467` from `MSC service thread 1-6` 96 milliseconds after the camel context started at `09:05:18,371` inside `MSC service thread 1-3`. This is quite a fast machine.

Do you think we should address this somehow?

Comment (jamesnetherton) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: It is a change in behaviour compared to how things work now.

My example is a bit of an edge case in some respects. However, if the Camel context state == Started, then I'd expect all of the relevant endpoints to be fully up and running at that point.

It's a tricky problem. Not sure if we want to put significant effort into resolving it now. We could create an issue and address it later.

@tdiesler what do you think?

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ppalaga) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: There indeed is a setting like this, but i do not want all chained method calls to be wrapped. E.g. the statement `addUndertowDomain.get("http-authentication-factory").set(HTTP_AUTH_FACTORY)` above looks OK unwrapped.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java (hunk 2 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: +1 create an issue and address it later

File: cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdiesler) [reply cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java]: The deployment suffix is a possible perhaps even likely indicator of j2ee deployment type, but not a required one I believe. Most ARQ deployments don't bother to use a suffix.

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tdiesler) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: Yes, please no such IDE hacks in the code.

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java (hunk 2 above)
Comment (tdiesler) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: I'd say, if it is not obviously wrong we can get it in and if there should be a problem with it in the future, we deal with it then.

File: cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppalaga) [reply cxfhttp/src/main/java/org/apache/cxf/transport/undertow/UndertowHTTPDestination.java]: > The deployment suffix is a possible perhaps even likely indicator of j2ee deployment type, but not a required one I believe.

Not required, if one does not want to rely on any http related functionality. WAR deployments are handled differently based on their .war archive name suffix inside WF and we depend on that functionality for CXF endpoints. Hence we may and have to check for the .war suffix of the class loader.

> Most ARQ deployments don't bother to use a suffix.

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    private static UserManager newUserManager() throws IOException {
+        final Path userPropertiesPath = WILDFLY_HOME.resolve("standalone/configuration/" + USERS_PROPS);
+        final Path rolePropertiesPath = WILDFLY_HOME.resolve("standalone/configuration/" + ROLES_PROPS);
+        return new UserManager(userPropertiesPath, rolePropertiesPath, SECURITY_REALM);
+    }
+
+    @Override
+    public void tearDown(ManagementClient managementClient, String containerId) throws Exception {
+        try (UserManager um = newUserManager()) {
+            um //
```
Comment (tdiesler) [inline itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: Why again would I want these funky line endings?

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ppalaga) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: I hope we all agree that

```
DMRUtils.batchNode() //
                .addStep(addRealm) //
                .addStep(addElytronDomain) //
                .addStep(addHttpAuthentication) //
                .addStep(addUndertowDomain) //
                .execute(client) //
                .assertSuccess();
```

is better readable than

```
DMRUtils.batchNode().addStep(addRealm).addStep(addElytronDomain).addStep(addHttpAuthentication)
                .addStep(addUndertowDomain).execute(client).assertSuccess();
```

So what is the point?

**Comment by tdiesler** [comment]
This is a massively complex piece of work. If it works only nearly as advertised, you deserve a medal. I'd say, lets bring it in and give it a collective spin when in master.

**Comment by ppalaga** [comment]
> lets bring it in and give it a collective spin when in master

We cannot merge to master before we get a camel release containg https://github.com/apache/camel/commit/829d8b6cb4935a9c4258e1c0642266cbe666c6d0

File: subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java (hunk 2 above)
Comment (ppalaga) [reply subsystem/core/src/main/java/org/wildfly/extension/camel/parser/CamelSubsystemAdd.java]: Filed https://github.com/wildfly-extras/wildfly-camel/issues/2600

File: itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesnetherton) [reply itests/standalone/basic/src/test/java/org/wildfly/camel/test/common/security/BasicSecurityDomainASetup.java]: I've never worked on a codebase where contributors need to hack their syntax, just to satisfy the flaws of some IDE or other tooling.

**Comment by ppalaga** [comment]
Nobody protested against the `jar -> war` commit 1d2829d so I am going to squash it.

As for the remaining open points:

* I'd prefer to keep the trailing `//` for a better readability
* "our own HTTP client" is not an issue, I hope
* check the context vs CXF endpoint init timing - has a separate issue https://github.com/wildfly-extras/wildfly-camel/issues/2600
* where in the outer class should inner classes be located? - I am ready to put them elsewhere but I heart no clear statement where is their proper place. End of outer class maybe?
* wildfly-camel-examples checked

**Comment by ppalaga** [comment]
In 98e75cc :

* removed trailing `//`
* "our own HTTP client" considered not an issue (unchanged)
* inner classes at the end of the outer class

All issues resolved from my PoV.

**Comment by ppalaga** [comment]
4cc52c2 removes the dump of the EAR to /home/ppalaga

**Pull request merged**